### PR TITLE
Refactor loggers code

### PR DESCRIPTION
**Pull request by AlessioQuercia**
## What does this PR do?
This PR serves as a base for PR #9187 and makes the necessary code refactor to the loggers to solve issue #9138 as follows:

- base.py:
  - Add `_scan_and_log_checkpoints` in LightningLoggerBase
  - Add `_log_checkpoints` in LightningLoggerBase
  - Add `scan_checkpoints`

- wandb.py:
  - Removed `_scan_and_log_checkpoints` from WandbLogger
  - Add `_log_checkpoints` in WandbLogger

In this way, `_scan_and_log_checkpoints` can be called by any logger, and in particular by MLFlowLogger (#9187). 

I am not sure if all the loggers should inherit `_scan_and_log_checkpoints` and `_log_checkpoints` or if it's better to define these functions only in the loggers that need it (WandbLogger and MLFlowLogger).

### Does your PR introduce any breaking changes? If yes, please list them.
None.

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

It was fun coding this 😃

### CODE REVIEW

**Comment by AlessioQuercia** [comment]
Any help here?

**Comment by tchaton** [comment]
> Any help here?

The PR seems a good way, is it ready for review ?

Best,
T.C

**Comment by AlessioQuercia** [comment]
> 
> 
> > Any help here?
> 
> The PR seems a good way, is it ready for review ?
> 
> Best,
> T.C

Yes, it was already somehow discussed in #9187. I am not sure if it's the best way of implementing it though. Plus, there are still 5 failing checks that I don't know how to solve.

Best regards,
Alessio

**Comment by carmocca** [comment]
@daniellepintz can you take a look at this PR?

**Comment by carmocca** [comment]
> I am not sure if all the loggers should inherit _scan_and_log_checkpoints and _log_checkpoints or if it's better to define these functions only in the loggers that need it (WandbLogger and MLFlowLogger).

We prefer the latter at this point. Mainly because these hooks would be internally called by the logger implementations and not by the `Trainer` or `LightningModule`

So we wouldn't want to do:

> Add _scan_and_log_checkpoints in LightningLoggerBase
Add _log_checkpoints in LightningLoggerBase

But the utility `scan_checkpoints` could be extracted as proposed.

I know this PR is quite old. But we'd still approve it if you still want to move it forward. The MLFlow logger is just generally unmaintained by its authors.

**Comment by Borda** [comment]
@AlessioQuercia @carmocca, how is it going here? :otter:

**Comment by AlessioQuercia** [comment]
Hi, I am very busy lately, I will try to update the PR towards the end of this month

**Comment by AlessioQuercia** [comment]
> But the utility `scan_checkpoints` could be extracted as proposed.

Should I move the utility to utilities/loggers.py and call it `_scan_checkpoints` instead @carmocca?

**Comment by carmocca** [comment]
Yes! :)

File: src/pytorch_lightning/loggers/logger.py
Comment (carmocca) [inline src/pytorch_lightning/loggers/logger.py:63]: This probably comes from a bad merge commit. Can you remove it?

**Comment by AlessioQuercia** [comment]
It should be ready now @carmocca @Borda.

Once this is accepted and merged, I can update PR [https://github.com/Lightning-AI/lightning/pull/9187](https://github.com/Lightning-AI/lightning/pull/9187).

File: tests/tests_pytorch/loggers/test_logger.py
Comment (carmocca) [inline tests/tests_pytorch/loggers/test_logger.py:327]: Can you re-write this test so that it doesn't need to train a model? You can instead mock the checkpoint_callback passed to `_scan_checkpoints`

Comment (AlessioQuercia) [reply tests/tests_pytorch/loggers/test_logger.py]: I re-wrote it, even though I am not sure it's exactly what you meant, as I am not familiar with mock. As it is now, no models are trained, just temporary files are created (I think that is needed by `_scan_checkpoints`).

**Review by awaelchli — APPROVED** [review]

**Review by carmocca — APPROVED** [review]

**Pull request merged**
